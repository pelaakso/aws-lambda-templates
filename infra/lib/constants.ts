export const LambdaMemory = {
  ONE_VCPU: 1769, // Approximately one vCPU
  HALF_VCPU: 892, // Approximately half of a vCPU
  QUARTER_VCPU: 512,
  TWO_VCPU: 3538,
} as const;
