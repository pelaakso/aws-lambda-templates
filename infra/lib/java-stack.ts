import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface Props extends StackProps {
  /**
   * Is this stack being created as part of a unit test?
   * If so, a dummy code asset is used so that real code assets don't need to exist.
   *
   * @default - It is assumed that stack is not created as part of unit test and real asset path is used.
   */
  readonly isUnitTest?: boolean;
}

export class JavaStack extends Stack {
  constructor(scope: Construct, id: string, props?: Props) {
    super(scope, id, props);

    new lambda.Function(this, 'BlankJava', {
      code: props?.isUnitTest
        ? lambda.Code.fromAsset('./test/resources/dummy-code.zip')
        : lambda.Code.fromAsset('../java/blank-java/target/blank-java-0.0.1-SNAPSHOT-package.zip'),
      handler: 'be.petey952.blankjava.Handler',
      runtime: lambda.Runtime.JAVA_11,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 512,
      timeout: Duration.seconds(15),
      functionName: 'blank-java-template-mvn',
      description: 'Blank Lambda template using Java, Maven build',
    });

    new lambda.Function(this, 'BlankJavaWithPowertools', {
      code: props?.isUnitTest
        ? lambda.Code.fromAsset('./test/resources/dummy-code.zip')
        : lambda.Code.fromAsset(
            '../java/blank-java-with-powertools/target/blank-java-with-powertools-0.0.1-SNAPSHOT-package.zip'
          ),
      handler: 'be.petey952.blankjavapowertools.Handler',
      runtime: lambda.Runtime.JAVA_11,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 512,
      timeout: Duration.seconds(15),
      environment: {
        POWERTOOLS_SERVICE_NAME: 'BlankJavaTemplateWithPowertools',
      },
      functionName: 'blank-java-template-with-powertools-mvn',
      description: 'Blank Lambda template using Java with Powertools, Maven build',
    });

    new lambda.Function(this, 'BlankJavaGradle', {
      code: props?.isUnitTest
        ? lambda.Code.fromAsset('./test/resources/dummy-code.zip')
        : lambda.Code.fromAsset('../java/blank-java/build/distributions/blank-java-0.0.1-SNAPSHOT-package.zip'),
      handler: 'be.petey952.blankjava.Handler',
      runtime: lambda.Runtime.JAVA_11,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 512,
      timeout: Duration.seconds(15),
      functionName: 'blank-java-template-gradle',
      description: 'Blank Lambda template using Java, Gradle build',
    });

    new lambda.Function(this, 'BlankJavaWithPowertoolsGradle', {
      code: props?.isUnitTest
        ? lambda.Code.fromAsset('./test/resources/dummy-code.zip')
        : lambda.Code.fromAsset(
            '../java/blank-java-with-powertools/build/distributions/blank-java-with-powertools-0.0.1-SNAPSHOT-package.zip'
          ),
      handler: 'be.petey952.blankjavapowertools.Handler',
      runtime: lambda.Runtime.JAVA_11,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 512,
      timeout: Duration.seconds(15),
      environment: {
        POWERTOOLS_SERVICE_NAME: 'BlankJavaTemplateWithPowertoolsGradle',
      },
      functionName: 'blank-java-template-with-powertools-gradle',
      description: 'Blank Lambda template using Java with Powertools, Gradle build',
    });
  }
}
