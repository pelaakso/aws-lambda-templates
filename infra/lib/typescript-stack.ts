import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as nodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { LambdaMemory } from './constants';

export interface Props extends StackProps {
  /**
   * Is this stack being created as part of a unit test?
   * If so, a dummy code asset is used so that real code assets don't need to exist.
   *
   * @default - It is assumed that stack is not created as part of unit test and real asset path is used.
   */
  readonly isUnitTest?: boolean;
}

export class TypescriptStack extends Stack {
  constructor(scope: Construct, id: string, props?: Props) {
    super(scope, id, props);

    new nodejs.NodejsFunction(this, 'BlankTypescript', {
      awsSdkConnectionReuse: true,
      entry: './typescript/blank-typescript/handler.ts',
      handler: 'handler',
      functionName: 'blank-typescript-template',
      description: 'Blank Lambda template using Typescript',
      architecture: lambda.Architecture.ARM_64,
      runtime: lambda.Runtime.NODEJS_22_X,
      memorySize: LambdaMemory.QUARTER_VCPU,
      timeout: Duration.seconds(15),
      loggingFormat: lambda.LoggingFormat.TEXT,
    });
  }
}
