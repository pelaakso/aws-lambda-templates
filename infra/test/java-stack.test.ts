import { Template } from 'aws-cdk-lib/assertions';
import * as cdk from 'aws-cdk-lib';
import * as Java from '../lib/java-stack';
import { expect, test } from 'vitest';

test('Java stack snapshot', () => {
  const stack = new cdk.Stack();

  const myStack = new Java.JavaStack(stack, 'MyTestStack', {
    env: { account: '140966923789', region: 'eu-central-1' },
    isUnitTest: true,
  });

  const template = Template.fromStack(myStack);

  // Replace all 64 byte asset hashes, because they are likely to change from build to build.
  // For example 02eaccf2c7a5bca24a1360de04a6ec227dfbebb07d930a867f0fe8ee5fc32f4d.zip
  expect.addSnapshotSerializer({
    test: (val) => (typeof val === 'string' && val.match(/[0-9a-f]{64}.zip/) ? true : false),
    print: (val) => {
      if (typeof val === 'string') {
        return val.replace(/[0-9a-f]{64}.zip/, '64-byte-asset-hash-removed.zip');
      }
      return `${val}`;
    },
  });

  expect(template).toMatchSnapshot();
});
