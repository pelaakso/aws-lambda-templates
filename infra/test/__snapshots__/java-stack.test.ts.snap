// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Java stack snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BlankJavaA787BB9F": Object {
      "DependsOn": Array [
        "BlankJavaServiceRole6CD6549F",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-140966923789-eu-central-1",
          "S3Key": "da2c20ecfd844e0cc630ce77e6932612a5f2dd5cba7e31348c8d0ead103126b1.zip",
        },
        "Description": "Blank Lambda template using Java",
        "FunctionName": "blank-java-template",
        "Handler": "be.petey952.blankjava.Handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "BlankJavaServiceRole6CD6549F",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BlankJavaServiceRole6CD6549F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BlankJavaWithPowertoolsCEDCD5E4": Object {
      "DependsOn": Array [
        "BlankJavaWithPowertoolsServiceRoleC5418871",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-140966923789-eu-central-1",
          "S3Key": "4267e2492d53303936ba3c2d2dfa48f05143516ba44707c064331b6f10df49b7.zip",
        },
        "Description": "Blank Lambda template using Java with Powertools",
        "Environment": Object {
          "Variables": Object {
            "POWERTOOLS_SERVICE_NAME": "BlankJavaTemplateWithPowertools",
          },
        },
        "FunctionName": "blank-java-template-with-powertools",
        "Handler": "be.petey952.blankjavapowertools.Handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "BlankJavaWithPowertoolsServiceRoleC5418871",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BlankJavaWithPowertoolsServiceRoleC5418871": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
